Sources/XCTest/Public/Asynchronous/XCTWaiter.swift

```diff
@@ -356,21 +349,19 @@ open class XCTWaiter {
 
 private extension XCTWaiter {
     func primitiveWait(using runLoop: RunLoop, duration timeout: TimeInterval) {
-        // The contract for `primitiveWait(for:)` explicitly allows waiting for a shorter period than requested
-        // by the `timeout` argument. Only run for a short time in case `cancelPrimitiveWait()` was called and
-        // issued `CFRunLoopStop` just before we reach this point.
-        let timeIntervalToRun = min(0.1, timeout)
-
-        // RunLoop.run(mode:before:) should have @discardableResult <rdar://problem/45371901>
-        _ = runLoop.run(mode: .default, before: Date(timeIntervalSinceNow: timeIntervalToRun))
+        
+        self.runLoopSource = RunLoop._Source()
+        self.runLoop?._add(self.runLoopSource!, forMode: .default)
+        _ = self.runLoop?.run(until: .init(timeIntervalSinceNow: timeout))
     }
 
     func cancelPrimitiveWait() {
-        guard let runLoop = runLoop else { return }
+        
 #if os(Windows)
+        guard let runLoop = runLoop else { return }
         runLoop._stop()
```

Comment:
Is this not equal to `runLoop?._stop()`?  I think that the right thing to do here is to remove the `#if os(Windows)` instead and have all platforms use `runLoopSource?.invalidate()`.  The reason for the original divergence is that `CFRunLoopStop` is not available on Windows as it is a CoreFoundation API.  This path is now using the SPI for the run loop source.

Reply:
@compnerd  have made the changes